### PR TITLE
Add API for discovering assets

### PR DESCRIPTION
**Pull request by andrewkolos**
Supersedes #116621. This PR is largely the same.
## Overview

This change introduces a public API for inspecting what assets are available within a Flutter app. Currently, there is no officially-supported way to do this, and some apps--needing this ability--depend on undocumented behavior internal to Flutter.

### Example use case
The [google_fonts](https://pub.dev/packages/google_fonts) package, when loading a font, first checks if the font is already available in the app. If it is not available, it will load it from fonts.google.com. [Code](https://github.com/material-foundation/google-fonts-flutter/blob/b32645187201388764f6d102579433d40eb1ad0b/lib/src/google_fonts_base.dart#L285).

Other known use cases:
* Preemptively searching for available shaders and then warming them up via a call over a MethodChannel.
* Dynamically listing assets (and maybe loading) all assets that match some criterion (e.g. are within some arbitrry directory).
* Tests.

### Terminology

**The asset manifest.** The framework, when performing it's first asset load, has to load a file called the asset manifest, which contains information about assets available to the app (e.g. where to find higher-resolution variants of images). This is a JSON file that is created at build time by the tool. It is meant for framework-internal use--but some Flutter apps, including ones internal to Google, read the file directly.

### Why make this change now

#117233 introduced a new, binary-formatted, version of the asset manifest that enables faster first-time asset resolution. Before updating the framework to use the new asset manifest and deprecating the older JSON manifest[^1], we should take this opportunity to add this new API.

## Design

We add a new class, `AssetManifest`, that is responsible for loading and parsing the asset manifest file contents from an `AssetBundle`. `AssetManifest` exposes two methods, one for listing all assets and one for listing the variants of a particular asset.

`AssetBundle` also gets a new method, `loadStructuredBinaryData`, to enable loading data from a binary asset. This new method is similar to the existing `loadStructuredData`, which is used for loading the JSON asset manifest.

[^1]:  Normally, it might be fair to perform such a breaking change--without a deprecation cycle--because the asset manifest is not a publicly documented/supported concept. However, it's possible that some users, when consulting with Flutter maintainers, were told about the file and that they could use it to accomplish the asset-related behavior they were looking for. As a result, it would be unfair to break the older behavior without a full deprecation cycle.
[^2]: It is difficult to use a type other than `Object?` for the map entry value type unless we commit to the idea that values only ever represent list of image variants.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 1]
```diff
@@ -225,16 +251,69 @@ abstract class CachingAssetBundle extends AssetBundle {
...
+        result = SynchronousFuture<T>(value);
+        if (completer != null) {
+          // The load and parse operation ran asynchronously. We already returned
+          // from the loadStructuredBinaryData function and therefore the caller
+          // was given the future of the completer.
+          completer.complete(value);
+        }
+      }, onError: (Object err, StackTrace? stack) {
+        completer!.completeError(err, stack);
+      });
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_bundle.dart:289]: you might be able to write this using a tear-off:

```suggestion
      }, onError: completer!.completeError);
```

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  /// Lists the keys of all known assets, not including asset variants.
+  Iterable<String> listAssets() {
+    throw UnimplementedError();
+  }
+
+  /// Gets available variants of an asset.
+  Iterable<AssetVariant> getAssetVariants(String key) {
+    throw UnimplementedError();
+  }
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:32]: These should be abstract methods instead of throwing, like:

```
 Iterable<String> listAssets();
  ```

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+/// The asset manifest is a map of asset files to a list of objects containing
+/// information about variants of that asset.
+///
+/// The entries with each variant object are:
+///  - "asset": the location of this variant to load it from.
+///  - "dpr": The device-pixel-ratio that the asset is best-suited for.
+///
+/// New fields could be added to this object schema to support new asset variation
+/// features, such as themes, locale/region support, reading directions, and so on.
+class _AssetManifestBin implements AssetManifest {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:46]: These docs won't be visible since the class itself is private, maybe there is somewhere else more appropriate to host it?

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+/// Contains information about an asset that is a variant of another asset.
+class AssetVariant {
+  /// Creates an object containing information about an asset variant.
+  AssetVariant({
+    required this.key,
+    required this.targetDevicePixelRatio,
+  });
+
+  /// The device pixel ratio that the asset is most ideal for, if any.
+  final double? targetDevicePixelRatio;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:89]: This would be null for any asset not in a 1x/2x subfolder right?

I would consider linking to the website assets page too.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  @override
+  Iterable<String> listAssets() {
+    return <String>[..._data.keys.cast<String>(), ..._typeCastedData.keys];
+  }
+}
+
+/// Contains information about an asset that is a variant of another asset.
+class AssetVariant {
+  /// Creates an object containing information about an asset variant.
+  AssetVariant({
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:83]:
```suggestion
  const AssetVariant({
```

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @override
+  Iterable<String> listAssets() {
+    return <String>[..._data.keys.cast<String>(), ..._typeCastedData.keys];
+  }
+}
+
+/// Contains information about an asset that is a variant of another asset.
+class AssetVariant {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:81]:
```suggestion
@immutable
class AssetVariant {
```

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    return loadFromAssetBundle(rootBundle);
+  }
+
+  /// Lists the keys of all known assets, not including asset variants.
+  Iterable<String> listAssets() {
+    throw UnimplementedError();
+  }
+
+  /// Gets available variants of an asset.
+  Iterable<AssetVariant> getAssetVariants(String key) {
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart:30]: I guess someone could modify it and it would mess with the cache though 🤷‍♂️

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 4 above)
Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:89]: > This would be null for any asset not in a 1x/2x subfolder right?

Yup, exactly. I chose to be passive and not doc comment that specifically--just in case things change in the future. A link to the doc site seems like a good idea, though.

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:32]: > I guess someone could modify it and it would mess with the cache though 🤷‍♂️

Yeah. I think we don't have readonly lists out-of-the-box in Dart. I suppose we could return a new copy of the list for every call--I imagine that, for the majority of use-cases, it wouldn't be terribly expensive in speed or memory. I still prefer returning `Iterable` whenever I have any doubts, but I don't know if this is a norm in the Flutter API.

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart:32]: No, generally speaking we return a List. if its small copying is a fine choice

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart:92]: Not documenting how an API behaves is not a reasonable defense for changing the API in the future. If we need to change the behavior, we have criteria on the wiki for determining whether or not it qualifies as a breaking change.

Please document the behavior

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:100]: If by "API" you are referring to how the ratio is derived from the file path, then I see your point. I made the doc string more informative and added a docsite link.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  /// Creates an object containing information about an asset variant.
+  AssetVariant({
+    required this.key,
+    required this.targetDevicePixelRatio,
+  });
+
+  /// The device pixel ratio that the asset is most ideal for, if any.
+  final double? targetDevicePixelRatio;
+
+  /// The asset's key. This can also be thought of as the logical name of an asset,
```
[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart]: 👍 

Removed the phrase.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 9]
```diff
@@ -0,0 +1,104 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../foundation.dart';
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:5]: since this is in services it should import through flutter/foundation and not a relative import.

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart:46]: If its not intended to be a doc comment use `//`

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  /// asset will be interpreted as 3.
+  ///
+  /// This will be null if the parent folder name is not a ratio value followed
+  /// by an "x".
+  ///
+  /// See [Declaring resolution-aware image assets](https://docs.flutter.dev/development/ui/assets-and-images#resolution-aware)
+  /// for more information.
+  final double? targetDevicePixelRatio;
+
+  /// The asset's key, which typically resembles a file location.
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:102]: I believe that specifically this is the URI encoded relative file URI, correct?

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:102]: IIRC the answer is yes for user-provided assets (i.e. ones in the current package). However, for assets provided by other packages it would look a little different (e.g. `packages/some_library/asset.png`). I would need to double-check all of this. I think I will do that and then make the doc string more specific.

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 11]
```diff
@@ -225,16 +251,67 @@ abstract class CachingAssetBundle extends AssetBundle {
...
+  /// and return the function's result.
+  ///
+  /// The result of parsing the bytedata is cached (the bytedata itself is not).
+  /// For any given `key`, the `parser` is only run the first time.
+  ///
+  /// Once the value has been parsed, the future returned by this function for
+  /// subsequent calls will be a [SynchronousFuture], which resolves its
+  /// callback synchronously.
+  @override
+  Future<T> loadStructuredBinaryData<T>(String key, FutureOr<T> Function(ByteData data) parser) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_bundle.dart:264]: Can you add some unit tests for this new API?

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 10 above)
Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart]: I added some more information to the docstring of `AssetManifest.listAssets`. Not sure if I should or should not add to the same to `AssetVariant.key`. I didn't talk about the package library case, though that is still discoverable through docsite links.

File: packages/flutter/lib/src/services/asset_bundle.dart (hunk 11 above)
Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_bundle.dart:264]: I added some tests for the caching behavior and also for a basic load. Let me know if there is any other behavior you would like to see unit tested.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  /// [AssetManifest] object from that data.
+  static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    return bundle.loadStructuredBinaryData(_kAssetManifestFilename, _AssetManifestBin.fromStandardMessageCodecMessage);
+  }
+
+  /// Loads asset manifest data from the root bundle and creates an
+  /// [AssetManifest] from that data.
+  static Future<AssetManifest> loadFromRootBundle() {
+    return loadFromAssetBundle(rootBundle);
+  }
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:26]: nit: I would remove this API and just use the loadFromAssetBundle with rootBundle

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+  /// The logical key maps to the path of an asset specified in the pubspec.yaml
+  /// file at build time.
+  ///
+  /// See [Specifying assets](https://docs.flutter.dev/development/ui/assets-and-images#specifying-assets)
+  /// and [Loading assets](https://docs.flutter.dev/development/ui/assets-and-images#loading-assets) for more
+  /// information.
+  List<String> listAssets();
+
+  /// Gets available variants of an asset.
+  List<AssetVariant> getAssetVariants(String key);
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:39]: Does this include a variant for the main asset?

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:33]: No it does not. I updated the doc string. However, this did make me realize something--this method should probably throw if `key` is not found within the manifest. Otherwise, it could be possible for users to mistakenly think that some asset has no variants when in reality they may be using some invalid hard-coded key. Do you agree?

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:33]: Thinking about this more, I want to get back to your original point. 

I think it also might be more intuitive from a user standpoint that an asset is not considered to be a variant of itself. However, it does make iterating over an asset and its variants more awkward. Also, it makes migration of code that consumes `AssetManifest.json` more error-prone since folks will have to remember to also operate on the main asset in addition to whatever loop they are replacing. I don't have a particular strong opinion on this at the very moment

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart:33]: I agree with this, but I am having trouble figuring out an API design that I think will be intuitive to users.

Specifically, the [Assets and images](https://docs.flutter.dev/development/ui/assets-and-images) doc makes a distinction between a _main_ asset and its _variants_:

> Then both graphics/background.png and graphics/dark/background.png are included in your asset bundle. The former is considered the main asset, while the latter is considered a variant.

Therefore, I think having `getAssetVariants` include the main asset in its results conflicts with this distinction. I have a few ideas:
1. Update the doc to say that each main asset is also considered to be a variant of itself--but, again, I don't know if people would generally find that intuitive.
2. The doc string on `getAssetVariants` can mention that, specifically in the context of this method, the main asset is considered to be a variant of itself.
3. Rename the `AssetVariant` class to something more broad, like `AssetMetadata`, and add a method `AssetManifest.getAssetMetadata(String key)` for getting the metadata for an asset without having to find it indirectly through `getAssetVariants`.

I'm biased towards option 2 since its simple and I don't see the API being popular enough to worry so much over, but I am very much interested in any thoughts you have.

Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart:33]: I like a combination of 2/3. AssetMetadata sounds like a better name, and we can document what the return is. At any rate, there needs to be a way to get the data for the main asset if you decide to use it.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+  factory _AssetManifestBin.fromStandardMessageCodecMessage(ByteData message) {
+    final dynamic data = const StandardMessageCodec().decodeMessage(message);
+    return _AssetManifestBin(data as Map<Object?, Object?>);
+  }
+
+  final Map<Object?, Object?> _data;
+  final Map<String, List<AssetMetadata>> _typeCastedData = <String, List<AssetMetadata>>{};
+
+  @override
+  List<AssetMetadata> describeAssetAndVariants(String key) {
```
Comment (andrewkolos) [inline packages/flutter/lib/src/services/asset_manifest.dart:64]: As discussed offline, I have this now include the main asset in the returned list.

I also renamed this from `getAssetVariants` to `describeAssetAndVariants`. I think this name is a bit long-winded but reduces ambiguity. Alternatively, we can update the docsite to specify that each asset is considered to be a variant of itself.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 15]
```diff
@@ -104,4 +121,11 @@ class AssetVariant {
   /// The asset's key, which is the path to the asset specified in the pubspec.yaml
   /// file at build time.
   final String key;
+
+  /// Whether or not this is a main asset. In other words, this is true if
+  /// this asset is not a variant of another.
+  ///
+  /// [Asset variants](https://docs.flutter.dev/development/ui/assets-and-images#asset-variants)
+  /// for more about asset variants.
+  final bool isMainAsset;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:130]: In general, we try to give properties fairly simple names and rely on that name + the type name to give full context. For example, instead of naming a property on Semantics `bool isSemanticsButton` we'd just call it `bool button`. In this case, I think the asset part is implied by the type being called `AssetMetadata` so you could shorten this to `main` and add more context via the doc comment.

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 16]
```diff
@@ -54,24 +58,36 @@ class _AssetManifestBin implements AssetManifest {
   }
 
   final Map<Object?, Object?> _data;
-  final Map<String, List<AssetVariant>> _typeCastedData = <String, List<AssetVariant>>{};
+  final Map<String, List<AssetMetadata>> _typeCastedData = <String, List<AssetMetadata>>{};
 
   @override
-  List<AssetVariant> getAssetVariants(String key) {
+  List<AssetMetadata> describeAssetAndVariants(String key) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_manifest.dart:64]: imo this is a bit long of a name.

Comment (andrewkolos) [reply packages/flutter/lib/src/services/asset_manifest.dart]: Alternatives:

* `getAssetVariants` (original)
* `describeAssetVariants` (`describe` is slightly more precise than `get` but is perhaps unnecessary since the return type already precisely indicates what the caller is getting back.
* `getAssetAndVariants`
* `describeAssetAndVariants` (long-winded but has the least potential for confusion imo)

Any other ideas?

Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_manifest.dart]: getAssetVariants is fine.


Explainaing exactly what the API does is a job for a doc comment, not the name of the method.

**Review by jonahwilliams — APPROVED** [review]
overall LGTM

**Comment by christopherfujino** [comment]
Note this was reverted in https://github.com/flutter/flutter/pull/119273 because of a compilation error post-submit
